Added Matter bridge application

kkasperczyk-no opened the pull request.

NordicBuilder commented:
## Test specification

### [CI/Jenkins/NRF](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-11880/1/)
- Integration Platforms

### [CI/Jenkins/integration](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-11880/1/)
| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-fw-nrfconnect-chip | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-11880/1/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

wiba-nordic reviewed (changes requested):
Reviewing the documentation in place of Greg.

I noticed there was a TODO section, so I guess this is still a work in progress, but I looked through the sections that looked finished.

kkasperczyk-no commented on applications/matter_bridge/README.rst:
```diff
@@ -0,0 +1,290 @@
+   This means that the sample has automatically started the Bluetooth LE advertising.
+#. Commission the device to the Matter network.
+   See `Commissioning the device`_ for more information.
+   During the commissioning process, write down the value for the bridge node ID, the temperature sensor node ID that is going to be used in the next steps (*<bridge_node_ID>*)
```
It's a copy paste error. There should be only bridge node ID.

kkasperczyk-no commented:
> Reviewing the documentation in place of Greg.
> 
> I noticed there was a TODO section, so I guess this is still a work in progress, but I looked through the sections that looked finished.

Thanks @wiba-nordic! I addressed your suggestions and filled missing TODO section (see last commit).

Damian-Nordic commented on applications/matter_bridge/Kconfig:
```diff
@@ -0,0 +1,47 @@
+
+endif
+
+source "${ZEPHYR_BASE}/../modules/lib/matter/config/nrfconnect/chip-module/Kconfig.features"
```
I think other apps already switched to using ZEPHYR_CONNECTEDHOMEIP_MODULE_DIR directly.

wiba-nordic commented on applications/matter_bridge/README.rst:
```diff
@@ -264,7 +271,25 @@ If you are new to Matter, the recommended approach is to use :ref:`CHIP Tool for
+When you start the commissioning procedure, the controller must get the onboarding information from the Matter accessory device.
+The onboarding information representation depends on your commissioner setup.
+
+For this sample, you can use one of the following :ref:`onboarding information formats <ug_matter_network_topologies_commissioning_onboarding_formats>` to provide the commissioner with the data payload that includes the device discriminator and the setup PIN code:
```
```suggestion
For this application, you can use one of the following :ref:`onboarding information formats <ug_matter_network_topologies_commissioning_onboarding_formats>` to provide the commissioner with the data payload that includes the device discriminator and the setup PIN code:
```

wiba-nordic reviewed (changes requested):
One very small change, and I just realized there's no release note entry. Seeing as it's a new application, one is warranted.

greg-fer commented on applications/matter_bridge/README.rst:
```diff
@@ -0,0 +1,315 @@
+
+You can use the following commands to control the bridge device:
+
+::
```
```suggestion
.. code-block:: console
```

greg-fer commented on applications/matter_bridge/README.rst:
```diff
@@ -0,0 +1,315 @@
+
+   uart:~$ matter_bridge add 256 "Kitchen Light"
+
+::
```
```suggestion
.. code-block:: console
```

wiba-nordic commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -228,6 +228,11 @@ Thingy:53: Matter weather station
+Matter Bridge
+-------------
+
+* Added :ref:`Matter bridge <matter_bridge_app>` application.
```
```suggestion
* Added the :ref:`Matter bridge <matter_bridge_app>` application.
```

greg-fer commented on applications/matter_bridge/README.rst:
```diff
@@ -0,0 +1,315 @@
+Matter CLI commands
+===================
+
+You can use the following commands to control the bridge device:
+
```
In the doc output, it will be hard to distinguish information about the two commands. I suggest you add lines 108-130 in a bullet point, and lines 132-146 in another. You can have:
```suggestion
You can use the following commands to control the bridge device:

* For adding a new bridged device to the bridge:

...

* For removes a new bridged device from the bridge:

```

kkasperczyk-no replied:
Ok, I think it was like this before the review from @wiba-nordic, so can we have some agreement on that?

kkasperczyk-no replied:
Or maybe not, it wasn't the same. Ok I will introduce it :)

kkasperczyk-no replied:
Ok, done, I hope it is correct now.

Damian-Nordic commented on applications/matter_bridge/README.rst:
```diff
@@ -0,0 +1,315 @@
+.. parsed-literal::
+   :class: highlight
+
+   uart:~$ matter_bridge add 256 "Kitchen Light"
```
Add a line that displays the assigned endpoint ID so that it's clear how a user knows the endpoint ID for the "remove" command?

greg-fer commented on applications/matter_bridge/README.rst:
```diff
@@ -0,0 +1,315 @@
+
+   matter_bridge remove <bridged_device_endpoint_id>
+
+This command removes a new bridged device from the bridge.
```
I would drop "new" here, because this would suggest you can only remove the last added bridge.

greg-fer commented on applications/matter_bridge/README.rst:
```diff
@@ -0,0 +1,315 @@
+This command removes a new bridged device from the bridge.
+In this command:
+
+   * ``<bridged_device_endpoint_id>`` is the endpoint id of the bridged device to be removed
```
```suggestion
   * ``<bridged_device_endpoint_id>`` is the endpoint ID of the bridged device to be removed.
```

greg-fer reviewed (commented):
Couple of suggestions.

wiba-nordic commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -348,6 +353,9 @@ Matter samples
+
+* :ref:`matter_thermostat_sample`:
+
   * Added :ref:`Matter thermostat <matter_thermostat_sample>` sample.
```
```suggestion
  * Added the :ref:`Matter thermostat <matter_thermostat_sample>` sample.
```
This is existing text, but might as well fix it since we're editing a close-by line anyway.

wiba-nordic reviewed (commented):
Looks good to my eye, but I'll check the doc build before giving approval.

wiba-nordic commented on applications/matter_bridge/README.rst:
```diff
@@ -0,0 +1,327 @@
+   In this command:
+
+   * ``<device_type>`` is a Matter device type to use for the bridged device.
+   The argument is mandatory and accepts the following values:
```
```suggestion
     The argument is mandatory and accepts the following values:
```
Wrong indent breaks doc build.

markaj-nordic reviewed (approved):
The implementation was discussed f2f prior to this PR, LGTM.

wiba-nordic commented:
We need 
```
   ../../applications/matter_bridge/README
```
as the second entry in `applications.rst`.

markaj-nordic commented on samples/matter/common/src/bridge/bridged_device_data_provider.h:
```diff
@@ -0,0 +1,25 @@
+	using UpdateAttributeCallback = void (*)(BridgedDeviceDataProvider &dataProvider, chip::ClusterId clusterId,
+						 chip::AttributeId attributeId, void *data, size_t dataSize);
+
+	explicit BridgedDeviceDataProvider(UpdateAttributeCallback callback) { mUpdateAttributeCallback = callback; }
```
I realized we missed virtual destructor here resulting in memory leaks when destroying child objects. I can fix it the follow up PR however.

ArekBalysNordic replied:
Please fix it in your PR, since we should finish this PR to unblock you.

greg-fer reviewed (changes requested):
We also need to:
- Mention this new application in all lines that say to check out the Matter weather station as the only Matter application, for example here:
![image](https://github.com/nrfconnect/sdk-nrf/assets/41291385/83dab9fd-a621-49c1-84e0-35afe68b3f1f)
Check for `matter_weather_station_app` and update the following pages (these can just be updated with :ref:`Matter applications<applications>`): protocols/matter/index.rst, matter/tools.rst, samples/matter.rst
- Update hw_requirements.rst with the memory values for the new application.
- Update data_model.rst with entries for the new application (and thermostat sample).

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-11880
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

wiba-nordic commented on applications/matter_bridge/README.rst:
```diff
@@ -0,0 +1,327 @@
+         I: Adding OnOff Light bridged device
+         I: Added device to dynamic endpoint 3 (index=0)
+
+* For removes a new bridged device from the bridge:
```
```suggestion
* For removing a bridged device from the bridge:
```

wiba-nordic commented on doc/nrf/applications.rst:
```diff
@@ -23,3 +23,4 @@ You can also run them on different hardware like a generic development kit, but
    ../../applications/machine_learning/README
    ../../applications/matter_weather_station/README
    ../../applications/zigbee_weather_station/README
+   ../../applications/matter_bridge/README
```
Should be between connectivity bridge and asset track, so that they are in alphabetical order based on the page title.
![image](https://github.com/nrfconnect/sdk-nrf/assets/83582261/d108be06-ff9d-431e-be0a-3704fe396f4d)

ArekBalysNordic commented:
@wiba-nordic @greg-fer I've provided all changes. We decided to postpone generating a memory report until we have the final support for the bridge. Could you please review the last commit?

wiba-nordic reviewed (approved):
Everything looks good to me.

markaj-nordic commented on applications/matter_bridge/src/bridged_device_types/onoff_light_data_provider.cpp:
```diff
@@ -0,0 +1,74 @@
+	switch (attributeId) {
+	case Clusters::OnOff::Attributes::OnOff::Id: {
+		mOnOff = *buffer;
+		NotifyUpdateState(clusterId, attributeId, &mOnOff, sizeof(mOnOff));
```
Not a blocker for this PR but should we really call `NotifyUpdateState()` here? To me, the the current flow is as follows:
- the Matter attribute is altered
- we call `BridgeManager::HandleWrite` which results in updating the bridged device state AND synchronizing the data provider as well by issuing polymorphic `BridgedDeviceDataProvider::UpdateState`
- in `OnOffLightDataProvider::UpdateState` we call `NotifyUpdateState` which brings us back to `BridgeManager::HandleUpdate` resulting in Matter bridged device attribute update which essentially was a trigger of the change in the first place

ArekBalysNordic replied:
Do you suggest repairing it here or unblocking your PR and considering it as a follow-up?

markaj-nordic replied:
I think we need @kkasperczyk-no 's insight here. For me we can live with it for now and fix later if needed.

ArekBalysNordic replied:
Ok, leave it as it is currently, and make a follow-up since it is the base of the whole project. :)

greg-fer commented on doc/nrf/protocols/matter/overview/data_model.rst:
```diff
@@ -303,3 +303,17 @@ HVAC device types
 | Fan               | A Fan device capable of controlling a fan in a heating or cooling system.    | Provisional         |                                       |
 +-------------------+------------------------------------------------------------------------------+---------------------+---------------------------------------+
+
+.. _ug_matter_device_types_lighting:
```
```suggestion
.. _ug_matter_device_types_utility:
```

greg-fer commented on applications/matter_bridge/README.rst:
```diff
@@ -0,0 +1,327 @@
+.. _matter_bridge_app:
+
+Matter: Bridge
+################
```
```suggestion
Matter bridge
##############
```
I would treat this on-par with "Matter controller" and other network topology devices.
See https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/protocols/matter/overview/network_topologies.html#matter-network-concepts.
Also because, assuming that "Matter:" is here to indicate the protocol and the second part is the device type, this could be interpreted we're adding an application for controlling bridges. ;)

greg-fer reviewed (changes requested):
Really sorry, two more changes.

greg-fer reviewed (approved):
Ship it! 🚀
